doc/mgmt/gnmi/gNMI_Subscription_for_YangData.md

```diff
@@ -0,0 +1,1203 @@
+  Translib will assume default values if app did not specify the preferences.
+  Preferences includes:
+  - Whether ON_CHANGE enabled
+  - Minimum SAMPLE interval.
+  - Preferred subscription mode during TARGET_DEFINED
+
+- Transformer common_app will discover the mappings and preferences from the annotations.
+  Subtree transformers should implement `Subscribe_xxx()` callback function to specify
+  the mappings and preferences.
+  Table and key transformers should handle wildcard paths.
+
+- App modules should implement `processSubscribe()` app interface function to resolve YANG keys for a DB entry.
+  This is required to get a specific YANG path from a wildcard path template.
+
+- Transformer common_app will automatically support this for simple cases.
+  Apps should implement a *path transformer* to handle special cases --
+  subtree transformer or when there is no one-to-one mapping of YANG key to DB key.
+
+### 2.2 Identifying YANG Based Subscriptions
+
+Today the gNMI server supports subscription based on multiple SONiC specific schemas -- like "event path", "DB paths" and "non-DB paths"
+as described in the [GRPC Telemetry HLD](../../system-telemetry/grpc_telemetry.md).
+It uses the request `target` attribute to identify type of the subscribe paths.
+A subscribe request will be rejected if `target` was not specified.
+
+This is not in sync with the [gNMI specification](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#2221-path-target).
+`target` is optional and is supposed to be an opaque data for the server.
+Hence, unlike existing steaming modules, no new reserved `target` value will be added for YANG paths.
+Instead, request `origin` based identification will be used, in accordance with the [gNMI Mixed Schema Specification](https://github.com/openconfig/reference/blob/master/rpc/gnmi/mixed-schema.md).
+Subscribe paths will be treated as translib YANG paths when the `origin` attribute is set as `openconfig`.
+For the sake of backward compatibility, the existing `target` based identification will be retained when the `origin` is not specified.
+Following table summarizes the behavior:
+
+| `origin` value  | `target` value  | Path type                                 |
+|-----------------|-----------------|-------------------------------------------|
+| openconfig      | *{opaque}*      | OpenConfig yang paths handled by translib |
+| *{empty}*       | CONFIG_DB<br>STATE_DB etc...  | DB path (existing logic)    |
+| *{empty}*       | EVENTS          | Event path (existing logic)               |
+| *{empty}*       | OTHERS          | Non-DB path (existing logic)              |
+| *{empty}*       | *{empty}*       | Error (existing logic)                    |
+| *{empty}*       | *{unknown}*     | Error                                     |
+| *{unknown}*     | *{opaque}*      | Error                                     |
+
```

Comment:
Please check https://github.com/sonic-net/SONiC/blob/master/doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md #1.2.1.3 Data Schema.
This table does not align with our decision tree, would you please update decision tree?

Reply:
I had referred that section earlier. That HLD only addresses schema identification for Get and Set rpcs. It does not talk about the Subscribe rpc. However, I saw few discrepancies in the implementation -- the Get & Set code is looking for `origin=sonic-db` only. I could not find any handling for `origin=sonic-yang` or `origin=sonic_yang` as described in the HLD.

The proposal in the current HLD is applicable for Subscribe rpc alone. It does not affect the Get and Set handling. Also, it is in alignment with decision matrix from the other HLD. It only looks different because it does not talk about origin values "sonic-db" and "sonic-yang". We do not support sonic yang based subscription as of now. Hence `origin=sonic-yang` is not applicable for the current HLD. Similarly, `origin=sonic-db` is not included because the existing code uses a `target` based path identification for DB, non-DB and events paths; and we cannot change it without breaking the backward compatibility.

Reply:
sonic-yang is not implemented for now.
For the Subscribe rpc, we might need to support sonic-db and sonic-yang. This requires a decision tree to handle different scenarios for this rpc. The existing code uses target and origin="", so supporting origin=sonic-db should not affect backward compatibility.

Reply:
And the logic of decision tree could be:
if origin == 'openconfig':
    # translib subscribe
elif origin == 'sonic-db':
    # sonic db subscribe
elif origin == 'sonic-yang':
    # sonic yang subscribe
else:
    # existing subscribe implementation

Reply:
As discussed in yesterday's meeting, the scope of this HLD is to support subscription for openconfig yang paths only. Hence I have not included anything about `origin=sonic-yang` in this HLD to avoid confusion. We do have plans to extend it to sonic ynag paths in a future release. Decision table will be updated at that time.

For sonic-db paths, the current implementation uses a different logic and we do not want to touch it -- for backward compatibility with existing clients. The path structure used also slightly differs from your HLD; please refer to [SONiC gRPC Telemetry](https://github.com/sonic-net/SONiC/blob/master/doc/system-telemetry/grpc_telemetry.md) HLD. There are other path types, like non-DB path, event path etc. Hence we are asking the owners of those components (I guess Microsoft) to enhance them.